ethereum/contracts/BasicInboundChannel.sol

```diff
@@ -17,10 +17,12 @@ contract BasicInboundChannel is InboundChannel {
     }
 
     // TODO: Submit should take in all inputs required for verification,
-    // including eg: _commitment, _parachainBlockNumber, _parachainMerkleProof, parachainHeadsMMRProof
-    function submit(Message[] memory _messages) public override {
-        //TODO: Verify messages
-        //verifyMessages(_messages, _commitment, ...);
+    // including eg: _parachainBlockNumber, _parachainMerkleProof, parachainHeadsMMRProof
+    function submit(Message[] memory _messages, bytes32 _commitment)
```

Comment:
maybe we can use calldata instead of memory for these fields to avoid them being copied over for each internal function call? (i'm not sure if we can easily access the calldata variables from internal calls - need to look into it)

^ let's leave this for a follow up PR/future optimization though if you don't have a quick and easy answer+fix - https://github.com/Snowfork/polkadot-ethereum/issues/252

Reply:
Looks like copying dynamically encoded structs from calldata to memory is supported as of [v0.7.6](https://github.com/ethereum/solidity/releases/tag/v0.7.6). I transitioned the inbound channels from calldata to memory in https://github.com/Snowfork/polkadot-ethereum/pull/266/commits/20a8833a91a4bfddd6037033ebf2944bdc9191f1.

Reply:
Yeh copying works but what I meant was to check if it is possible and cheaper to avoid copying by accessing the calldata directly within our internal functions instead of sending them as memory parameters.

Reply:
@musnit I think Denali did implement that (the `message` parameter is now passed via calldata in all internal functions)

Reply:
ah wait, I think I see what you mean - the `_message` parameter already lives in `msg.data` in ABI-encoded form. So we should just extract the bytes from there:

So something like this in pseudocode:
```
commitmentBytes= msg.data[START...END]
commitmentHash = keccak256(commitmentBytes) 
```

From https://ethereum.stackexchange.com/questions/14037/what-is-msg-data

Reply:
actually even better, can extract the data using the `CALLDATACOPY` instruction (described in https://blog.openzeppelin.com/ethereum-in-depth-part-2-6339cf6bddb9)

Reply:
@vgeddes he's referring to copying each message into memory inside the `processMessages` function and then operating upon it instead of just reading directly from the calldata. Just pushed the related updates in https://github.com/Snowfork/polkadot-ethereum/pull/266/commits/0ce021b7cb499d5c43e67f3130feb8a9334dabc3. The strategy does save us a bit of gas:
- Before: `Submitted 4 messages. Gas used: 154248`
- After: `Submitted 4 messages. Gas used: 153816`

Reply:
Looks like it saves exactly 108 gas per message 👍

Reply:
ahh OK! yeah that makes sense.

Reply:
looks good, thanks!